spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,89 +113,251 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Output.class)) {
+				outputAnnotationCount++;
+			}
+		}
+		return outputAnnotationCount;
+	}
+
+	private boolean isDeclarativeStreamListenerMethod(Method method, String methodAnnotatedinboundName, String methodAnnotatedOutboundName) {
+		int methodArgumentsLength = method.getParameterTypes().length;
+		for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Input.class)) {
+				String inboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Input.class));
+				Assert.isTrue(StringUtils.hasText(inboundName), "@Input annotation should always be associated with a valid inbound name");
+				return isDeclarativeMethodParameter(getBindableBean(inboundName), methodParameter);
+			}
+			else if (methodParameter.hasParameterAnnotation(Output.class)) {
+				String outboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Output.class));
+				Assert.isTrue(StringUtils.hasText(outboundName), "@Output annotation should always be associated with a valid outbound name");
+				return isDeclarativeMethodParameter(getBindableBean(outboundName), methodParameter);
+			}
+			else if (StringUtils.hasText(methodAnnotatedOutboundName)) {
+				return isDeclarativeMethodParameter(getBindableBean(methodAnnotatedOutboundName), methodParameter);
+			}
+			else if (StringUtils.hasText(methodAnnotatedinboundName)) {
+				return isDeclarativeMethodParameter(getBindableBean(methodAnnotatedinboundName), methodParameter);
+			}
+		}
+		return false;
+	}
+
+	private boolean isDeclarativeMethodParameter(Object targetBean, MethodParameter methodParameter) {
+		if (targetBean != null) {
+			if (methodParameter.getParameterType().isAssignableFrom(targetBean.getClass())) {
+				return true;
+			}
+			for (StreamListenerParameterAdapter<?, Object> streamListenerParameterAdapter : streamListenerParameterAdapters) {
+				if (streamListenerParameterAdapter.supports(targetBean.getClass(), methodParameter)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	private String getInboundElementNameFromMethodAnnotation(StreamListener streamListener) {
+		return StringUtils.hasText(streamListener.value()) ? streamListener.value() : null;
+	}
+
+	private String getOutboundElementNameFromMethodAnnotation(Method method) {
+		SendTo sendTo = AnnotationUtils.findAnnotation(method, SendTo.class);
+		if (sendTo != null) {
+			Assert.isTrue(!ObjectUtils.isEmpty(sendTo.value()), "At least one output must be specified");
+			Assert.isTrue(sendTo.value().length == 1, "Multiple destinations cannot be specified");
```

Comment:
@ilayaperumalg if we were to support multiple values (which I don't think we should), it would be in any case semantically different from using different parameters ('broadcast to all specified destinations' which is not what the distinct parameters imply)